### PR TITLE
Ruby 2.7: Beginless ranges fully supported

### PR DESCRIPTION
**Pull request by LillianZ**
Finishing up from https://github.com/oracle/truffleruby/pull/2155

In this PR, added support for:
    Range#{cover?, include?, ===}.
    String#{byteslice, slice, slice!} and Symbol#slice 
    Kernel#{caller, caller_locations} and Thread#backtrace_locations

https://github.com/Shopify/truffleruby/issues/1

### CODE REVIEW

File: src/main/ruby/truffleruby/core/kernel.rb
Comment (eregon) [inline src/main/ruby/truffleruby/core/kernel.rb:730]: Could you put the parens around the arguments like `Primitive.nil?(start.begin)`?
I think that's a lot more common, and parens around a condition are rare in Ruby, unless they need to affect grouping of conditions like (a || b) && c.

**Comment by eregon** [comment]
I'll let @norswap review the rest.

File: src/main/ruby/truffleruby/core/truffle/range_operations.rb
Comment (norswap) [inline src/main/ruby/truffleruby/core/truffle/range_operations.rb:85]: This mostly isn't your doing, but I find this `cover_range?` logic incredibly difficult to read for something that should essentially be very simple.

Is there any reason we couldn't rewrite the whole thing as:

```ruby
def self.range_cover?(range, other)
  b1 = range.begin
  b2 = other.begin
  e1 = range.end
  e2 = other.end

  return false if Primitive.nil?(b2) && !Primitive.nil?(b1)
  return false if Primitive.nil?(e2) && !Primitive.nil?(e1)

  (Primitive.nil?(b2) || cover?(range, b2)) && (Primitive.nil?(e2) || cover?(range, e2))
end
```

File: src/main/ruby/truffleruby/core/truffle/range_operations.rb
Comment (norswap) [inline src/main/ruby/truffleruby/core/truffle/range_operations.rb:116]: Same as `cover_range?`, hard to follow because of the existing code.

Proposal (completely untested):
```ruby
def self.cover?(range, value)
  # Check lower bound.
  if !Primitive.nil?(range.begin)
    # MRI uses <=> to compare, so must we.
    cmp = (range.begin <=> value)
    return false if Compare.compare_int(cmp) > 0
  end

  # Check upper bound.
  if !Primitive.nil(range.end)
    value = value + (range.exclude_end? 1 : 0)
    cmp = (value <=> range.end)
    return false if Comparable.compare_int(cmp) > 0
  end

  true
end
```

**Review by norswap — COMMENTED** [review]
Looks good. Great work on all these new specs 👍

File: src/main/ruby/truffleruby/core/truffle/range_operations.rb
Comment (LillianZ) [reply src/main/ruby/truffleruby/core/truffle/range_operations.rb]: That looks so much better, thank you! Slight modifications due to exclude_end? things.

File: src/main/ruby/truffleruby/core/truffle/range_operations.rb
Comment (LillianZ) [reply src/main/ruby/truffleruby/core/truffle/range_operations.rb:97]: Also much more beautiful, thanks!

Also, `r_less` is no longer being used, so I commented it out; not sure if it should be removed altogether.

Comment (eregon) [reply src/main/ruby/truffleruby/core/truffle/range_operations.rb:97]: Yes, let's remove `range_less` if it's unused :)
(we can always use the git history to restore it if it's useful in the future)

**Review by norswap — APPROVED** [review]
Seems good to go for me, unless there are more changes that need to go in?

**Comment by LillianZ** [comment]
I'm done and ready to merge. Thank you for the great suggestions!

**Comment by chrisseaton** [comment]
Great PR to finish a tricky set of specs!

**Comment by eregon** [comment]
There has some CI failures, I'm not sure why GitHub Actions didn't run on the last changes.
In MRI tests:
```

  1) Failure:
Bug::Test_ExceptionAT#test_exception_at_throwing [/b/b/e/main/test/mri/tests/cext-ruby/exception/test_exception_at_throwing.rb:7]:
assert_separately failed with error message
pid 22634 exit 0
| resource:/truffleruby/core/truffle/range_operations.rb:93: warning: `-' after local variable or literal is interpreted as binary operator
| resource:/truffleruby/core/truffle/range_operations.rb:93: warning: even though it seems like unary operator


  2) Failure:
OpenSSL::TestEngine#test_engines_free [/b/b/e/main/test/mri/tests/openssl/test_engine.rb:76]:
assert_separately failed with error message
pid 22838 exit 0
| resource:/truffleruby/core/truffle/range_operations.rb:93: warning: `-' after local variable or literal is interpreted as binary operator
| resource:/truffleruby/core/truffle/range_operations.rb:93: warning: even though it seems like unary operator
```

Lint:
```
Offenses:

src/main/ruby/truffleruby/core/truffle/range_operations.rb:85:68: C: Layout/TrailingWhitespace: Trailing whitespace detected.
      return false unless (Primitive.nil?(b2) || cover?(range, b2)) 
                                                                   ^
```

**Comment by LillianZ** [comment]
Thanks for letting me know, I think it's fixed now! 

The non-running tests could have something to do with the fact the CHANGELOG is conflicting? In the future I'll wait until after the reviews to write it.

**Comment by eregon** [comment]
> The non-running tests could have something to do with the fact the CHANGELOG is conflicting?

That seems likely, I didn't think about it.
Something worth mentioning in that GitHub issue about an incorrectly-reported conflict, cc @chrisseaton 

A workaround would be to rebase the PR, but we'll try first if this now passes the internal CI.

**Pull request merged**
